rustls-mbedcrypto-provider/src/fips_utils.rs

```diff
@@ -0,0 +1,256 @@
+struct FipsCheckError(mbedtls::Error);
+
+impl fmt::Display for FipsCheckError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl std::error::Error for FipsCheckError {}
+
+impl From<FipsCheckError> for rustls::Error {
+    fn from(value: FipsCheckError) -> Self {
+        OtherError(Arc::new(value)).into()
+    }
+}
+
+/// Wrap a [`mbedtls::Error`] with [`FipsCheckError`] and convert it to [`rustls::Error`].
+#[inline]
+fn wrap_mbedtls_error_as_fips(mbed_err: mbedtls::Error) -> rustls::Error {
+    FipsCheckError(mbed_err).into()
+}
+
+/// ECC Full Public-Key Validation Routine, which is defined in
+/// section 5.6.2.3.3 of [NIST SP 800-56A Rev. 3]
+///
+/// [NIST SP 800-56A Rev. 3]:
+///     https://csrc.nist.gov/pubs/sp/800/56/a/r3/final
+pub(crate) fn fips_check_ec_pub_key(ec_mbed_pk: &Pk) -> Result<(), rustls::Error> {
+    let mut rng = crate::rng::rng_new().ok_or(rustls::Error::FailedToGetRandomBytes)?;
+    fips_check_ec_pub_key_mbed(ec_mbed_pk, &mut rng).map_err(wrap_mbedtls_error_as_fips)?;
+    log::info!("ECC Full Public-Key Validation: passed");
+    Ok(())
+}
+
+/// ECC Pairwise Consistency Test upon generation that mimics the shared
+/// secret computation using the recent generated key pair and a known
+/// key pair with the same domain parameters and comparing the shared
+/// secret computation values calculated using either FFC DH or ECC CDH
+/// primitives.
+///
+/// Learn more in:
+/// - [FIPS 140-3 IG] section 10.3.A.
+/// - [SP 800-56Ar3] section 5.6.2.1.4.
+///
+/// [FIPS 140-3 IG]: https://csrc.nist.gov/projects/cryptographic-module-validation-program/fips-140-3-ig-announcements
+/// [SP 800-56Ar3]: https://csrc.nist.gov/pubs/sp/800/56/a/r3/final
+pub(crate) fn fips_ec_pct(ec_mbed_pk: &mut Pk, ec_group_id: EcGroupId) -> Result<(), rustls::Error> {
+    let mut rng = crate::rng::rng_new().ok_or(rustls::Error::FailedToGetRandomBytes)?;
+    // Get a static ec pub key based on given [`EcGroupId`]
+    let known_ec_key = get_known_ec_key(&ec_group_id).expect("validated");
+    let mut known_ec_key = known_ec_key
+        .lock()
+        .map_err(|_| rustls::Error::General("Failed to get known ec key: poisoned lock".to_string()))?;
+
+    fips_ec_pct_mbed(ec_mbed_pk, &mut known_ec_key, &mut rng).map_err(wrap_mbedtls_error_as_fips)?;
+    log::info!("ECC Pairwise Consistency Test: passed");
+    Ok(())
+}
+
+#[allow(non_snake_case)]
+fn fips_ec_pct_mbed<F: mbedtls::rng::Random>(
+    ec_mbed_pk: &mut Pk,
+    known_ec_key: &mut Pk,
+    rng: &mut F,
+) -> Result<(), mbedtls::Error> {
+    // Pairwise Consistency Test upon generation that mimics the
+    // shared secret computation using the recent generated key
+    // pair and a known key pair with the same domain parameters
+    // and comparing the shared secret computation values
+    // calculated using either FFC DH or ECC CDH primitives.
+    let mut shared_1 = vec![0; ECDSA_MAX_LEN];
```

Comment:
MbedTLS is overly conservative with `ECDSA_MAX_LEN`. And also here it does not really make sense to refer to ECDSA, but ECDH.

The elliptic-curve Diffie-Hellman exchange produces a shared secret which is the x coordinate of a point in the curve, so for prime curves (such as NIST) we have that the max length is p which you can access by `EcGroup::p()` followed by `byte_length()`

Reply:
Got it!

Reply:
fixed